### PR TITLE
Torque

### PR DESCRIPTION
**Pull request by davidmanzanares**
Partially adds #16 

This PR adds a new `torque` expression to create animated maps in a more straightforward manner.

### CODE REVIEW

File: src/core/renderer.js [hunk 1]
```diff
@@ -5,6 +5,8 @@ import { Asc, Desc } from './style/functions';
 
 const HISTOGRAM_BUCKETS = 1000;
 
+const initialTimestamp = Date.now();
```
Comment (IagoLast) [inline src/core/renderer.js]: why here and not in the constructor?

🌈  uppercase and snake case for global constants

File: src/core/style/expressions/expression.js [hunk 2]
```diff
@@ -61,6 +61,14 @@ export default class Expression {
         return this._getChildren().map(child => child._getDrawMetadataRequirements()).reduce(schema.union, schema.IDENTITY);
     }
 
+    /**
+     * Pre-rendering routine. Should establish the current timestamp in seconds since an arbitrary point in time as needed.
+     * @param {Number} timestamp
+     */
+    _setTimestamp(timestamp) {
```
Comment (IagoLast) [inline src/core/style/expressions/expression.js]: 🌈 if this method is public used from the renderer) why the underscore prefix?

File: src/core/style/expressions/expression.js [hunk 3]
```diff
@@ -61,6 +61,14 @@ export default class Expression {
         return this._getChildren().map(child => child._getDrawMetadataRequirements()).reduce(schema.union, schema.IDENTITY);
     }
 
+    /**
+     * Pre-rendering routine. Should establish the current timestamp in seconds since an arbitrary point in time as needed.
+     * @param {Number} timestamp
```
Comment (IagoLast) [inline src/core/style/expressions/expression.js]: `Number` --> `number`

File: src/core/style/expressions/torque.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+ * Create an animated temporal filter (torque)
+ *
+ * @param {carto.style.expression.expression} input input to base the temporal filter,
+ * if input is a property, the beginning and end of the animation will be determined by the minimum and maximum timestamps of the property on the dataset,
+ * this can be problematic if outliers are present. Otherwise input must be a float expression in which 0 means beginning of the animation and 1 means end.
+ * @param {Number} duration duration of the animation in seconds
+ * @param {carto.style.expression.fade} fade fadeIn/fadeOut configuration
+ * @return {carto.style.expressions.torque}
+ *
+ * @example `width:    2
```
Comment (IagoLast) [inline src/core/style/expressions/torque.js]: 🌈  wrong indentation

File: src/core/style/expressions/torque.js [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        const input = this.input.eval(feature);
+        const cycle = this._cycle.eval(feature);
+        const duration = this.duration;
+        const fadeIn = this.fade.fadeIn.eval(feature);
+        const fadeOut = this.fade.fadeOut.eval(feature);
+        return 1 - clamp(Math.abs(input - cycle) * duration / (input > cycle ? fadeIn : fadeOut), 0, 1);
+    }
+}
+
+function clamp(x, min, max) {
```
Comment (IagoLast) [inline src/core/style/expressions/torque.js]: I think the clamp function could be moved to `utils` since is widely used

File: src/core/style/expressions/utils.js [hunk 6]
```diff
@@ -1,6 +1,8 @@
 import { float, category } from '../functions';
 import Expression from './expression';
 
+export const DEFAULT = undefined;
```
Comment (IagoLast) [inline src/core/style/expressions/utils.js]: ?

File: src/core/style/expressions/viewportAggregation.js [hunk 7]
```diff
@@ -40,7 +40,11 @@ function generateAggregattion(metadataPropertyName, global) {
             return this.property._getMinimumNeededSchema();
         }
         _getDrawMetadataRequirements() {
-            return { columns: [this.property.name] };
+            if (!global) {
```
Comment (IagoLast) [inline src/core/style/expressions/viewportAggregation.js]: is this related to te issue?

File: src/core/style/expressions/torque.js [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+// fadeIn and fadeOut of 0.5 seconds
+filter:       torque($day, 40, fade(0.5))
+
+`
+ *
+ * @memberof carto.style.expressions
+ * @name fade
+ * @function
+*/
+export class Fade extends Expression {
```
Comment (IagoLast) [inline src/core/style/expressions/torque.js]: I think `Fade` should be in its own file

File: src/core/renderer.js (hunk 1 above)
Comment (davidmanzanares) [reply src/core/renderer.js]: why here and not in the constructor?

Because this is just needed for numeric stability, no need to have a per-instance constant

File: src/core/style/expressions/expression.js (hunk 2 above)
Comment (davidmanzanares) [reply src/core/style/expressions/expression.js]: Consistency with all the other expression methods that are public to the renderer, yet private to the user.

File: src/core/style/expressions/torque.js (hunk 4 above)
Comment (davidmanzanares) [reply src/core/style/expressions/torque.js]: In the example? How should it be?

File: src/core/style/expressions/torque.js (hunk 5 above)
Comment (davidmanzanares) [reply src/core/style/expressions/torque.js]: Totally agree

File: src/core/style/expressions/torque.js (hunk 4 above)
Comment (IagoLast) [reply src/core/style/expressions/torque.js]: - keep the `*` at the beginning of the lines
- the example starts in the next line (after @example tag) 
- there is no need for the backtick
-

File: src/core/style/expressions/utils.js (hunk 6 above)
Comment (davidmanzanares) [reply src/core/style/expressions/utils.js]: @rochoa and I talked about adding a `DEFAULT` expression to be more explicit in the API to be used like: `torque($day, default, fade(5))`. I think we should have it at least on the string API, I don't have much opinion regarding the JS API

File: src/core/style/expressions/viewportAggregation.js (hunk 7 above)
Comment (davidmanzanares) [reply src/core/style/expressions/viewportAggregation.js]: `torque()` triggers this function, and since it didn't check for `global` it made the animations slower. It is related, although the change would make sense without torque too.

File: src/core/style/expressions/torque.js (hunk 4 above)
Comment (IagoLast) [reply src/core/style/expressions/torque.js]:
```javascript
 * @example <caption> fadeIn of 0.1 seconds, fadeOut of 0.3 seconds </caption>
 * const s = carto.style.expressions;
 * new carto.Style({
 *  filter: s.torque($day, 40, s.fade(0.1, 0.3))
 * });
 * 
 * @example <caption>   fadeIn and fadeOut of 0.5 seconds </caption>
 * const s = carto.style.expressions;
 * new carto.Style({
 *  filter: s.torque($day, 40, s.fade(0.5))
 * });
```

File: src/core/style/expressions/torque.js (hunk 8 above)
Comment (davidmanzanares) [reply src/core/style/expressions/torque.js]: I haven't moved it since, at least right now, `Fade` is entirely made to be used with `torque`

File: src/core/style/expressions/expression.js (hunk 3 above)
Comment (davidmanzanares) [reply src/core/style/expressions/expression.js]: :+1:

File: src/core/style/expressions/torque.js (hunk 8 above)
Comment (davidmanzanares) [reply src/core/style/expressions/torque.js]: I haven't moved it since, at least right now, `Fade` is entirely made to be used with `torque`

File: src/core/style/expressions/torque.js (hunk 4 above)
Comment (davidmanzanares) [reply src/core/style/expressions/torque.js]: :+1: Thank you, updated
